File: intern/73.set-matrix-zeroes.js
/*
 * @lc app=leetcode id=73 lang=javascript
 *
 * [73] Set Matrix Zeroes
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
const setZeroes = function (matrix) {
    const m = matrix.length;
    if (m < 1) {
        return;
    }
    // m >= 1
    const n = matrix[0].length;
    if (n < 1) {
        return;
    }
    // m >= 1 && n >= 1
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (matrix[i][j] === 0) {
                for (let k = 0; k < m; k++) {
                    if (matrix[k][j] !== 0) {
                        matrix[k][j] = NaN;
                    }
                }
                for (let k = 0; k < n; k++) {
                    if (matrix[i][k] !== 0) {
                        matrix[i][k] = NaN;
                    }
                }
            }
        }
    }
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (Number.isNaN(matrix[i][j])) {
                matrix[i][j] = 0;
            }
        }
    }
};
// @lc code=end

File: src/2025/73.set-matrix-zeroes.cpp
/*
 * @lc app=leetcode id=73 lang=cpp
 *
 * [73] Set Matrix Zeroes
 */

#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  void setZeroes(std::vector<std::vector<int>>& matrix) {
    // Use the first element of row/col as indicators.
    // use matrix[0][0] to indicate matrix[0] is filled with 0.

    // Use the variable to indicate matrix[][0] is filled with 0.
    bool is_column_0_zero = false;
    for (int i = 0; i < matrix.size(); i++) {
      for (int j = 0; j < matrix[0].size(); j++) {
        if (matrix[i][j] == 0) {
          if (j == 0) {
            // If it is in first column, mark the variable instead of setting
            // matrix[0][0].
            is_column_0_zero = true;
          } else {
            matrix[i][0] = 0;
            matrix[0][j] = 0;
          }
        }
      }
    }

    for (int i = 1; i < matrix.size(); i++) {
      if (matrix[i][0] == 0) {
        for (int j = 0; j < matrix[0].size(); j++) {
          matrix[i][j] = 0;
        }
      }
    }

    for (int j = 1; j < matrix[0].size(); j++) {
      if (matrix[0][j] == 0) {
        for (int i = 0; i < matrix.size(); i++) {
          matrix[i][j] = 0;
        }
      }
    }

    // Set first rows/cols at last.

    if (matrix[0][0] == 0) {
      for (int j = 0; j < matrix[0].size(); j++) {
        matrix[0][j] = 0;
      }
    }

    if (is_column_0_zero) {
      for (int i = 0; i < matrix.size(); i++) {
        matrix[i][0] = 0;
      }
    }
  }
};
// @lc code=end
